GafferImage::Mix : Added Mix node

danieldresser opened the pull request:
Hopefully this is about what you're picturing - like a merge, except only two inputs, no operation to choose, and a mask plug.  It shares a lot of code with Merge, and for the tests I basically started with the merge tests and removed the ones that didn't make sense.

johnhaddon commented on include/GafferImage/Mix.h:
```diff
@@ -0,0 +1,104 @@
+namespace GafferImage
+{
+
+/// A node for Merging two or more images. Mix and metadata from the first input;
```
This comment seems to just be copy-pasted from Merge, and doesn't really apply here.

johnhaddon commented on include/GafferImage/Mix.h:
```diff
@@ -0,0 +1,104 @@
+		virtual Imath::Box2i computeDataWindow( const Gaffer::Context *context, const ImagePlug *parent ) const;
+		/// Creates a union of all of the connected inputs channelNames.
+		virtual IECore::ConstStringVectorDataPtr computeChannelNames( const Gaffer::Context *context, const ImagePlug *parent ) const;
+		/// Implemented to call doMix according to operationPlug()
```
This comment also seems inaccurate.

johnhaddon commented on python/GafferImageTest/MixTest.py:
```diff
@@ -0,0 +1,346 @@
+
+		mix = GafferImage.Mix()
+
+		expectedHash = r1["out"].image().hash()
```
This isn't testing what it claims to be testing. We want to be comparing the Gaffer hashes that will be used to share entries in Gaffer's cache (`r1["out"].imageHash()`), but instead we're computing the image and comparing the hash of the data we get. Since this is copy pasted from MergeTest, any chance you could fix that too?

johnhaddon commented on python/GafferImageUI/MixUI.py:
```diff
@@ -0,0 +1,91 @@
+			"""
+			The channel which controls the blend.  Clamped between 0 and 1.
+			0 to take first input, 1 to take second input.
+			""",
```
It seems like this would benefit from the [channel widget from ShuffleUI](https://github.com/GafferHQ/gaffer/blob/master/python/GafferImageUI/ShuffleUI.py#L145). Could you either take a look at doing that, or just add a todo mentioning the existing functionality in ShuffleUI?

johnhaddon commented on src/GafferImage/Mix.cpp:
```diff
@@ -0,0 +1,291 @@
+	:	ImageProcessor( name, 2, 2 )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+    addChild( new ImagePlug( "mask", Gaffer::Plug::In ) );
```
Spaces make the indentation fairies cry.

johnhaddon commented on src/GafferImage/Mix.cpp:
```diff
@@ -0,0 +1,291 @@
+	}
+
+	const std::string &maskChannel = maskChannelPlug()->getValue();
+	h.append( maskPlug()->channelDataHash( maskChannel, tileOrigin ) );
```
We need to check `ImageAlgo::channelExists( maskPlug(), maskChannel )` here before calling `channelDataHash()`.

johnhaddon replied:
Just to clarify - we get crashes when entering non-existent channels otherwise.

johnhaddon commented on src/GafferImage/Mix.cpp:
```diff
@@ -0,0 +1,291 @@
+	const Box2i tileBound( tileOrigin, tileOrigin + V2i( ImagePlug::tileSize() ) );
+
+	const std::string &maskChannel = maskChannelPlug()->getValue();
+	ConstFloatVectorDataPtr maskData = maskPlug()->channelData( maskChannel, tileOrigin );
```
Same here, we need to call `ImageAlgo::channelExists()`.

johnhaddon commented on src/GafferImageBindings/MixBinding.cpp:
```diff
@@ -0,0 +1,55 @@
+
+void bindMix()
+{
+	scope s = GafferBindings::DependencyNodeClass<Mix>();
```
No need for the scope - this was there in MergeBinding to scope the enum, but didn't need to be copy-pasted.

danieldresser commented:
OK, I think I've addressed all feedback.

Tricky bits for you to look at:
* the API for ChannelPlugValueWidget ( mostly just metadata naming )
* the merge test fixes:  fixing one test turned up a problem that was easy to fix, but the other seems like it could be a bit difficult, and I'm not sure that merges with one input hooked up are really an important case to put work into optimizing, so now that the tests are fixed, one of them is an expectedFailure.

johnhaddon commented on src/GafferImage/Mix.cpp:
```diff
@@ -0,0 +1,314 @@
+		h.append( validBound );
+	}
+
+	mixPlug()->hash( h );
```
I think we could definitely benefit from direct pass-throughs for the `mix==0` and `mix==1 && !maskPlug()->getInput()` cases. These will be essential if we use Mix to add a masking feature to every SceneProcessor. No need to add it immediately if it doesn't fit into your schedule, but would be nice to have (along with tests) if it does.

danieldresser commented:
Comments addressed.  Adding the pass-throughs felt like there was some potential to get things subtly wrong, but it appears to be working.

> Can you think of any other potential uses for extra channels? I'm wondering if we should just have allowBlack and allowWhite metadata instead?

I kept it the implementation and just updated the comment.  I can't think of other definite use cases, but it seems fairly plausible?  Maybe some node wants to support __luminance, or something like __matchingChannelInBaseLayer if you wanted to have the option of shuffling R/G/B to foo.R/foo.G/foo.B in one go or something.  Nothing crucial I can think of, I can do allowBlack/allowWhite if you prefer.

johnhaddon commented on src/GafferImage/Mix.cpp:
```diff
@@ -186,6 +230,21 @@ void Mix::hashChannelData( const GafferImage::ImagePlug *output, const Gaffer::C
+		return;
+	}
+
+	h.append( mix );
```
The `ImageProcessor::hashChannelData()` call should be here, just before the first `append()` - when we're making a direct pass through, there's no point initialising the hash because we're about to overwrite it.

johnhaddon commented on src/GafferImage/Mix.cpp:
```diff
@@ -186,6 +230,21 @@ void Mix::hashChannelData( const GafferImage::ImagePlug *output, const Gaffer::C
 {
 	ImageProcessor::hashChannelData( output, context, h );
 
+	float mix = mixPlug()->getValue();
```
I'm trying to get into the habit of making everything possible const, so the compiler will tell me when I make mistakes like `if( mix = 0.0f )`, and so there's less to think about when following the code flow. Would be nice to see this be const (and the others), but no big deal...

johnhaddon commented:
Thanks Daniel! Just the ordering of the `hashChannelData()` call to tweak now - could you do that, squash, and then merge?
Cheers...
John
